# Part I - Exploring the Prosper Loan Dataset
## By Sondos Aabed

> My Github: [@sondosaabed](https://github.com/sondosaabed) 

> My LinkedIn: [@sondosaabed](https://www.linkedin.com/in/sondosaabed/)

<hr/>

## Abstract


<hr/>

## Table of Contents
- Introduction
    - Data Dictionary
- Objectives
- Premirely Wrangling
- Univariate extrapolation
- Bivariate extrapolation
- Multivariate extrapolation
- Conclusions

<hr/>

## Introduction

> Introduce the dataset

### Data Dictionary
The following data dictionary shows each variable of the dataset and the corresponding description:

| Variable                                | Description                                                                                                                                                                   |
|-----------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ListingKey                              | Unique key for each listing, same value as the 'key' used in the listing object in the API.                                                                                   |
| ListingNumber                           | The number that uniquely identifies the listing to the public as displayed on the website.                                                                                    |
| ListingCreationDate                     | The date the listing was created.                                                                                                                                             |
| CreditGrade                             | The Credit rating that was assigned at the time the listing went live. Applicable for listings pre-2009 period and will only be populated for those listings.                 |
| Term                                    | The length of the loan expressed in months.                                                                                                                                   |
| LoanStatus                              | The current status of the loan: Cancelled, Chargedoff, Completed, Current, Defaulted, FinalPaymentInProgress, PastDue. The PastDue status will be accompanied by a delinquency bucket. |
| ClosedDate                              | Closed date is applicable for Cancelled, Completed, Chargedoff and Defaulted loan statuses.                                                                                   |
| BorrowerAPR                             | The Borrower's Annual Percentage Rate (APR) for the loan.                                                                                                                     |
| BorrowerRate                            | The Borrower's interest rate for this loan.                                                                                                                                   |
| LenderYield                             | The Lender yield on the loan. Lender yield is equal to the interest rate on the loan less the servicing fee.                                                                  |
| EstimatedEffectiveYield                 | Effective yield is equal to the borrower interest rate (i) minus the servicing fee rate, (ii) minus estimated uncollected interest on charge-offs, (iii) plus estimated collected late fees.  Applicable for loans originated after July 2009. |
| EstimatedLoss                           | Estimated loss is the estimated principal loss on charge-offs. Applicable for loans originated after July 2009.                                                                |
| EstimatedReturn                         | The estimated return assigned to the listing at the time it was created. Estimated return is the difference between the Estimated Effective Yield and the Estimated Loss Rate. Applicable for loans originated after July 2009. |
| ProsperRating (numeric)                 | The Prosper Rating assigned at the time the listing was created: 0 - N/A, 1 - HR, 2 - E, 3 - D, 4 - C, 5 - B, 6 - A, 7 - AA.  Applicable for loans originated after July 2009.  |
| ProsperRating (Alpha)                   | The Prosper Rating assigned at the time the listing was created between AA - HR.  Applicable for loans originated after July 2009.                                            |
| ProsperScore                            | A custom risk score built using historical Prosper data. The score ranges from 1-10, with 10 being the best, or lowest risk score.  Applicable for loans originated after July 2009. |
| ListingCategory                         | The category of the listing that the borrower selected when posting their listing: 0 - Not Available, 1 - Debt Consolidation, 2 - Home Improvement, 3 - Business, 4 - Personal Loan, 5 - Student Use, 6 - Auto, 7 - Other, 8 - Baby&Adoption, 9 - Boat, 10 - Cosmetic Procedure, 11 - Engagement Ring, 12 - Green Loans, 13 - Household Expenses, 14 - Large Purchases, 15 - Medical/Dental, 16 - Motorcycle, 17 - RV, 18 - Taxes, 19 - Vacation, 20 - Wedding Loans |
| BorrowerState                           | The two letter abbreviation of the state of the address of the borrower at the time the Listing was created.                                                                   |
| Occupation                              | The Occupation selected by the Borrower at the time they created the listing.                                                                                                 |
| EmploymentStatus                        | The employment status of the borrower at the time they posted the listing.                                                                                                    |
| EmploymentStatusDuration                | The length in months of the employment status at the time the listing was created.                                                                                            |
| IsBorrowerHomeowner                     | A Borrower will be classified as a homeowner if they have a mortgage on their credit profile or provide documentation confirming they are a homeowner.                        |
| CurrentlyInGroup                        | Specifies whether or not the Borrower was in a group at the time the listing was created.                                                                                     |
| GroupKey                                | The Key of the group in which the Borrower is a member of. Value will be null if the borrower does not have a group affiliation.                                              |
| DateCreditPulled                        | The date the credit profile was pulled.                                                                                                                                       |
| CreditScoreRangeLower                   | The lower value representing the range of the borrower's credit score as provided by a consumer credit rating agency.                                                         |
| CreditScoreRangeUpper                   | The upper value representing the range of the borrower's credit score as provided by a consumer credit rating agency.                                                         |
| FirstRecordedCreditLine                 | The date the first credit line was opened.                                                                                                                                    |
| CurrentCreditLines                      | Number of current credit lines at the time the credit profile was pulled.                                                                                                     |
| OpenCreditLines                         | Number of open credit lines at the time the credit profile was pulled.                                                                                                        |
| TotalCreditLinespast7years              | Number of credit lines in the past seven years at the time the credit profile was pulled.                                                                                     |
| OpenRevolvingAccounts                   | Number of open revolving accounts at the time the credit profile was pulled.                                                                                                  |
| OpenRevolvingMonthlyPayment             | Monthly payment on revolving accounts at the time the credit profile was pulled.                                                                                              |
| InquiriesLast6Months                    | Number of inquiries in the past six months at the time the credit profile was pulled.                                                                                         |
| TotalInquiries                          | Total number of inquiries at the time the credit profile was pulled.                                                                                                          |
| CurrentDelinquencies                    | Number of accounts delinquent at the time the credit profile was pulled.                                                                                                      |
| AmountDelinquent                        | Dollars delinquent at the time the credit profile was pulled.                                                                                                                 |
| DelinquenciesLast7Years                 | Number of delinquencies in the past 7 years at the time the credit profile was pulled.                                                                                        |
| PublicRecordsLast10Years                | Number of public records in the past 10 years at the time the credit profile was pulled.                                                                                      |
| PublicRecordsLast12Months               | Number of public records in the past 12 months at the time the credit profile was pulled.                                                                                     |
| RevolvingCreditBalance                  | Dollars of revolving credit at the time the credit profile was pulled.                                                                                                        |
| BankcardUtilization                     | The percentage of available revolving credit that is utilized at the time the credit profile was pulled.                                                                      |
| AvailableBankcardCredit                 | The total available credit via bank card at the time the credit profile was pulled.                                                                                           |
| TotalTrades                             | Number of trade lines ever opened at the time the credit profile was pulled.                                                                                                  |
| TradesNeverDelinquent                   | Number of trades that have never been delinquent at the time the credit profile was pulled.                                                                                   |
| TradesOpenedLast6Months                 | Number of trades opened in the last 6 months at the time the credit profile was pulled.                                                                                       |
| DebtToIncomeRatio                       | The debt to income ratio of the borrower at the time the credit profile was pulled. This value is Null if the debt to income ratio is not available. This value is capped at 10.01 (any debt to income ratio larger than 1000% will be returned as 1001%). |
| IncomeRange                             | The income range of the borrower at the time the listing was created.                                                                                                         |
| IncomeVerifiable                        | The borrower indicated they have the required documentation to support their income.                                                                                          |
| StatedMonthlyIncome                     | The monthly income the borrower stated at the time the listing was created.                                                                                                   |
| LoanKey                                 | Unique key for each loan. This is the same key that is used in the API.                                                                                                       |
| TotalProsperLoans                       | Number of Prosper loans the borrower at the time they created this listing. This value will be null if the borrower had no prior loans.                                       |
| TotalProsperPaymentsBilled              | Number of on time payments the borrower made on Prosper loans at the time they created this listing. This value will be null if the borrower had no prior loans.             |
| OnTimeProsperPayments                   | Number of on time payments the borrower had made on Prosper loans at the time they created this listing. This value will be null if the borrower has no prior loans.         |
| ProsperPaymentsLessThanOneMonthLate     | Number of payments the borrower made on Prosper loans that were less than one month late at the time they created this listing. This value will be null if the borrower had no prior loans. |
| ProsperPaymentsOneMonthPlusLate         | Number of payments the borrower made on Prosper loans that were greater than one month late at the time they created this listing. This value will be null if the borrower had no prior loans. |
| ProsperPrincipalBorrowed                | Total principal borrowed on Prosper loans at the time the listing was created. This value will be null if the borrower had no prior loans.                                     |
| ProsperPrincipalOutstanding             | Principal outstanding on Prosper loans at the time the listing was created. This value will be null if the borrower had no prior loans.                                        |
| ScorexChangeAtTimeOfListing             | Borrower's credit score change at the time the credit profile was pulled. This will be the change relative to the borrower's last Prosper loan. This value will be null if the borrower had no prior loans. |
| LoanCurrentDaysDelinquent               | The number of days delinquent.                                                                                                                                               |
| LoanFirstDefaultedCycleNumber           | The cycle the loan was charged off. If the loan has not charged off the value will be null.                                                                                   |
| LoanMonthsSinceOrigination              | Months since the loan originated.                                                                                                                                            |
| LoanNumber                              | The number that uniquely identifies the loan to the public as displayed on the website.                                                                                      |
| LoanOriginalAmount                      | The original amount of the loan.                                                                                                                                             |
| LoanOriginationDate                     | The date the loan originated.                                                                                                                                                |
| LoanOriginationQuarter                  | The quarter in which the loan originated.                                                                                                                                    |
| MemberKey                               | Unique key for each member. This is the same key that is used in the API.                                                                                                     |
| MonthlyLoanPayment                      | The monthly payment (principal and interest) the borrower is required to make for this loan.                                                                                 |
| LP_CustomerPayments                     | The total payments (principal + interest) that have been made on the loan by the borrower.                                                                                   |
| LP_CustomerPrincipalPayments            | The total principal payments that have been made on the loan by the borrower.                                                                                                |
| LP_InterestandFees                      | Interest and fees paid by the borrower.                                                                                                                                      |
| LP_ServiceFees                          | The servicing fees paid by the borrower.                                                                                                                                    |
| LP_CollectionFees                       | The collection fees paid by the borrower.                                                                                                                                   |
| LP_GrossPrincipalLoss                   | Gross principal loss on the loan.                                                                                                                                           |
| LP_NetPrincipalLoss                     | Net principal loss on the loan.                                                                                                                                             |
| LP_NonPrincipalRecoverypayments         | Non-principal recovery payments on the loan.                                                                                                                                |
| PercentFunded                           | The percentage of the loan that was funded.                                                                                                                                 |
| Recommendations                         | Number of recommendations the borrower had at the time they created the listing.                                                                                            |
| InvestmentFromFriendsCount              | Number of investments that were made by friends at the time the listing was created.                                                                                        |
| InvestmentFromFriendsAmount             | The dollar amount of investments that were made by friends at the time the listing was created.                                                                              |
| Investors                               | The number of investors that funded the loan.                                                                                                                               |


<hr/>

## Objectives

**1.  Loan Performance Analysis**

**2.  Credit Score and Borrower Analysis**

**3.  Geographic and Demographic Analysis**

<hr/>

## Preliminary Wrangling

- In this section, a preliminary data wrangling is done on the dataset. Here is what this include:
- 

In [1]:
## import all packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

c:\Users\SS\AppData\Local\Programs\Python\Python311\Lib\site-packages\scipy\__init__.py:169: UserWarning: A NumPy version >=1.18.5 and <1.26.0 is required for this version of SciPy (detected version 1.26.4
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


### Loading the dataset

> Let's Load in the dataset into a pandas dataframe:

In [2]:
df = pd.read_csv("./data/prosperLoanData.csv") ## Load the csv into pandas dataframe
df.sample(10) ## Looking at a random sample of 10 rows.

ListingKey  ListingNumber            ListingCreationDate  \
27092  CB9634336380590687DBE20         406422  2008-09-30 13:12:14.443000000   
14497  991935980102465886D06A4        1083924  2013-12-12 14:58:50.020000000   
7672   85FF3419444286167CCE467         326510  2008-05-05 17:34:24.690000000   
34817  D14434280416331043A198E         380200  2008-08-09 10:20:45.567000000   
54295  B3363534327177422F73143         545063  2011-12-17 09:35:08.843000000   
87869  461F351926224788948A8F6         514174            2011-07-03 10:27:08   
47501  1B5D3580253134016C47BD1         794480  2013-06-02 20:44:31.577000000   
52361  A76E35738499637675B6261         737730  2013-03-25 19:06:13.637000000   
98332  25A53578959897020C51A54         788900  2013-05-26 10:58:25.890000000   
25473  2BFA337978192389575203A          92943  2007-02-03 16:54:59.260000000   

      CreditGrade  Term  LoanStatus           ClosedDate  BorrowerAPR  \
27092           D    36   Completed  2011-10-08 00:00:00      0.23226   
14497         NaN    60     Current                  NaN      0.14658   
7672           HR    36  Chargedoff  2010-01-15 00:00:00      0.22237   
34817           D    36   Completed  2011-08-10 00:00:00      0.33383   
54295         NaN    36  Chargedoff  2013-10-29 00:00:00      0.34577   
87869         NaN    60     Current                  NaN      0.14657   
47501         NaN    60     Current                  NaN      0.22283   
52361         NaN    60     Current                  NaN      0.31293   
98332         NaN    60     Current                  NaN      0.13227   
25473          AA    36  Chargedoff  2009-01-15 00:00:00      0.11194   

       BorrowerRate  LenderYield  ...  LP_ServiceFees  LP_CollectionFees  \
27092        0.2000       0.1900  ...          -29.55             0.0000   
14497        0.1239       0.1139  ...          -33.77             0.0000   
7672         0.2000       0.1900  ...          -10.95             0.0000   
34817        0.3099       0.2999  ...          -54.25           -43.8600   
54295        0.3058       0.2958  ...          -20.66             0.0000   
87869        0.1329       0.1229  ...         -312.01             0.0000   
47501        0.1984       0.1884  ...         -129.46             0.0000   
52361        0.2862       0.2762  ...          -35.42             0.0000   
98332        0.1099       0.0999  ...         -156.17             0.0000   
25473        0.1050       0.0950  ...          -34.50           -87.7588   

       LP_GrossPrincipalLoss  LP_NetPrincipalLoss  \
27092                   0.00                 0.00   
14497                   0.00                 0.00   
7672                  706.63               706.63   
34817                   0.00                 0.00   
54295                 857.90               857.90   
87869                   0.00                 0.00   
47501                   0.00                 0.00   
52361                   0.00                 0.00   
98332                   0.00                 0.00   
25473                4519.48              4488.22   

      LP_NonPrincipalRecoverypayments  PercentFunded  Recommendations  \
27092                             0.0            1.0                0   
14497                             0.0            1.0                0   
7672                              0.0            1.0                0   
34817                             0.0            1.0                0   
54295                             0.0            1.0                0   
87869                             0.0            1.0                0   
47501                             0.0            1.0                0   
52361                             0.0            1.0                0   
98332                             0.0            1.0                0   
25473                             0.0            1.0                0   

      InvestmentFromFriendsCount InvestmentFromFriendsAmount Investors  
27092                          0     

### Dataset Structure

In [3]:
df.shape ## showing the shape of the dataset

(113937, 81)

> This dataset has 113,937 rows and 81 columns. Which is a relatively big dataset.

### Dataset Assessment and Cleaning

#### Duplicated Records

In [14]:
df.duplicated(subset='LoanKey').sum()

871

> Let's identify the duplicated records based on the LoanKey and see if wwe should handle this

In [24]:
duplicates = df[df.duplicated(subset='LoanKey', keep=False)]
display(duplicates.head(10))

ListingKey  ListingNumber            ListingCreationDate  \
8    0F043596202561788EA13D5        1023355  2013-12-02 10:43:39.117000000   
9    0F043596202561788EA13D5        1023355  2013-12-02 10:43:39.117000000   
29   0F563597161095613517437        1051243  2013-12-17 09:18:33.220000000   
176  106335993636414276CB477        1119836  2014-01-08 14:27:50.320000000   
313  09233589620788733CFB8CE         930842  2013-09-25 08:03:11.860000000   
349  313635901230654318A9030         931467  2013-09-26 18:50:29.053000000   
442  09AD35918712001025AC1BD         969821  2013-10-24 13:21:31.607000000   
444  09CD3592594126374FB0A7C         986199  2013-10-18 08:28:03.610000000   
455  31C73597152310464749E00        1092437  2013-12-23 13:47:35.500000000   
461  44F2358557406858060EBDE         870200  2013-08-15 07:12:49.410000000   

    CreditGrade  Term LoanStatus ClosedDate  BorrowerAPR  BorrowerRate  \
8           NaN    36    Current        NaN      0.07620        0.0629   
9           NaN    36    Current        NaN      0.07620        0.0629   
29          NaN    36    Current        NaN      0.15223        0.1239   
176         NaN    36    Current        NaN      0.32446        0.2850   
313         NaN    36    Current        NaN      0.19144        0.1550   
349         NaN    36    Current        NaN      0.17090        0.1349   
442         NaN    36    Current        NaN      0.20524        0.1685   
444         NaN    36    Current        NaN      0.22773        0.1905   
455         NaN    36    Current        NaN      0.17151        0.1355   
461         NaN    60    Current        NaN      0.18965        0.1660   

     LenderYield  ...  LP_ServiceFees  LP_CollectionFees  \
8         0.0529  ...          -16.77                0.0   
9         0.0529  ...          -16.77                0.0   
29        0.1139  ...          -29.73                0.0   
176       0.2750  ...           -3.40                0.0   
313       0.1450  ...          -36.97                0.0   
349       0.1249  ...          -15.40                0.0   
442       0.1585  ...           -8.41                0.0   
444       0.1805  ...          -42.03                0.0   
455       0.1255  ...           -6.40                0.0   
461       0.1560  ...          -49.69                0.0   

     LP_GrossPrincipalLoss  LP_NetPrincipalLoss  \
8                      0.0                  0.0   
9                      0.0                  0.0   
29                     0.0                  0.0   
176                    0.0                  0.0   
313                    0.0                  0.0   
349                    0.0                  0.0   
442                    0.0                  0.0   
444                    0.0                  0.0   
455                    0.0                  0.0   
461                    0.0                  0.0   

    LP_NonPrincipalRecoverypayments  PercentFunded  Recommendations  \
8                               0.0            1.0                0   
9                               0.0            1.0                0   
29                              0.0            1.0                0   
176                             0.0            1.0                0   
313                             0.0            1.0                0   
349                             0.0            1.0                0   
442                             0.0            1.0                0   
444                             0.0            1.0                0   
455                             0.0            1.0                0   
461                             0.0            1.0                0   

    InvestmentFromFriendsCount InvestmentFromFriendsAmount Investors  
8                            0                         0.0         1  
9                            0                         0.0         1  
29                           0                         0.0         1  
176                          0                         

In [25]:
df.drop_duplicates(subset='LoanKey', keep='first', inplace=True)

In [28]:
print(df.duplicated(subset='LoanKey').sum())
df.shape

0


(113066, 81)

> Let's check for the duplicated records based on the `ListingKey` based on teh documentation it has to be unique too.

In [29]:
df.duplicated(subset='ListingKey').sum()

0

#### Data types Validity

- Assessment: Let's look at the data types of these variables and assess them using `.info`:

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   ListingKey                           113937 non-null  object 
 1   ListingNumber                        113937 non-null  int64  
 2   ListingCreationDate                  113937 non-null  object 
 3   CreditGrade                          28953 non-null   object 
 4   Term                                 113937 non-null  int64  
 5   LoanStatus                           113937 non-null  object 
 6   ClosedDate                           55089 non-null   object 
 7   BorrowerAPR                          113912 non-null  float64
 8   BorrowerRate                         113937 non-null  float64
 9   LenderYield                          113937 non-null  float64
 10  EstimatedEffectiveYield              84853 non-null   float64
 11  EstimatedLoss

> We have the columns of `ClosedDate`, `LoanOriginationDate`, `DateCreditPulled`, and the `ListingCreationDate` has an object type and it has to be a datetime type.

In [5]:
## list the date columns
date_columns = ['ClosedDate', 'LoanOriginationDate', 'DateCreditPulled', 'ListingCreationDate'] 

### Loopong through the list and coverting to datetime data type
for col in date_columns:
    df[col] = pd.to_datetime(df[col], format='mixed') ## using the format as mixed to infer the format for each element individually

df.info() ### check if that is successful

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
 #   Column                               Non-Null Count   Dtype         
---  ------                               --------------   -----         
 0   ListingKey                           113937 non-null  object        
 1   ListingNumber                        113937 non-null  int64         
 2   ListingCreationDate                  113937 non-null  datetime64[ns]
 3   CreditGrade                          28953 non-null   object        
 4   Term                                 113937 non-null  int64         
 5   LoanStatus                           113937 non-null  object        
 6   ClosedDate                           55089 non-null   datetime64[ns]
 7   BorrowerAPR                          113912 non-null  float64       
 8   BorrowerRate                         113937 non-null  float64       
 9   LenderYield                          113937 non-null  float64       
 

> Now the rest of the data types of the variables are valid.

#### Data Completness

In [6]:
def get_percent_null(df):
    """
    Args:
        - df (pd.Dataframe)
    Returns:
        - The percentages of missing values
    """
    null_counts = df.isnull().sum()
    null_counts = null_counts[null_counts > 0].sort_values()
    return (null_counts/df.shape[0])*100

In [7]:
get_percent_null(df)

BorrowerAPR                             0.021942
CreditScoreRangeUpper                   0.518708
CreditScoreRangeLower                   0.518708
PublicRecordsLast10Years                0.611742
CurrentDelinquencies                    0.611742
InquiriesLast6Months                    0.611742
TotalCreditLinespast7years              0.611742
FirstRecordedCreditLine                 0.611742
DelinquenciesLast7Years                 0.868901
TotalInquiries                          1.017229
EmploymentStatus                        1.979164
Occupation                              3.149109
BorrowerState                           4.840394
AvailableBankcardCredit                 6.621203
TradesOpenedLast6Months                 6.621203
TradesNeverDelinquent (percentage)      6.621203
TotalTrades                             6.621203
CurrentCreditLines                      6.673864
OpenCreditLines                         6.673864
PublicRecordsLast12Months               6.673864
RevolvingCreditBalan

> There are many columns that has null values, some of which have a very high count of null values. The following columns were the highest percentage of null such as higher than 50%. These columns are droppped.

In [8]:
high_missing_percent = ['CreditGrade', 'ScorexChangeAtTimeOfListing', 'LoanFirstDefaultedCycleNumber', 'ClosedDate', 'TotalProsperLoans',
 'TotalProsperPaymentsBilled', 'OnTimeProsperPayments', 'ProsperPaymentsLessThanOneMonthLate', 'ProsperPaymentsOneMonthPlusLate', 
 'ProsperPrincipalBorrowed', 'ProsperPrincipalOutstanding']

In [9]:
df.drop(columns=high_missing_percent, inplace=True)

In [10]:
df.shape

(113937, 70)

> The drop is checked now that the variables went from 81 to 70.

Handling the other null values by dropping NA values and rows:

In [11]:
get_percent_null(df)

BorrowerAPR                            0.021942
CreditScoreRangeUpper                  0.518708
CreditScoreRangeLower                  0.518708
TotalCreditLinespast7years             0.611742
CurrentDelinquencies                   0.611742
FirstRecordedCreditLine                0.611742
PublicRecordsLast10Years               0.611742
InquiriesLast6Months                   0.611742
DelinquenciesLast7Years                0.868901
TotalInquiries                         1.017229
EmploymentStatus                       1.979164
Occupation                             3.149109
BorrowerState                          4.840394
AvailableBankcardCredit                6.621203
TradesOpenedLast6Months                6.621203
TotalTrades                            6.621203
TradesNeverDelinquent (percentage)     6.621203
PublicRecordsLast12Months              6.673864
RevolvingCreditBalance                 6.673864
BankcardUtilization                    6.673864
CurrentCreditLines                     6

> These columns are identifiers, mostly unique between each loan.They are irrelevant for the task and therefore they should be dropped too.

In [12]:
df.drop(columns=["ListingKey", "ListingNumber", "GroupKey", "LoanKey", "LoanNumber", "MemberKey", "DateCreditPulled"], inplace= True )

In [13]:
df.shape

(113937, 63)

> The drop is done, since the features count went down to 63.

In [14]:
get_percent_null(df)

BorrowerAPR                            0.021942
CreditScoreRangeUpper                  0.518708
CreditScoreRangeLower                  0.518708
InquiriesLast6Months                   0.611742
TotalCreditLinespast7years             0.611742
FirstRecordedCreditLine                0.611742
PublicRecordsLast10Years               0.611742
CurrentDelinquencies                   0.611742
DelinquenciesLast7Years                0.868901
TotalInquiries                         1.017229
EmploymentStatus                       1.979164
Occupation                             3.149109
BorrowerState                          4.840394
AvailableBankcardCredit                6.621203
TradesOpenedLast6Months                6.621203
TotalTrades                            6.621203
TradesNeverDelinquent (percentage)     6.621203
PublicRecordsLast12Months              6.673864
RevolvingCreditBalance                 6.673864
BankcardUtilization                    6.673864
OpenCreditLines                        6

> Let's handle the lower percentages by only dropping the NA values instead of the whole columns.

In [15]:
df.dropna(inplace=True)

In [16]:
get_percent_null(df)

Series([], dtype: float64)

> No missing NA values left.

In [17]:
df.shape

(76216, 63)

> After handling missing values and dropping unneccary columns the shape of the data is 76,216 rows and 63 columns

### Main Features of interest
- All the left columns are the features of intrest they will be devided based on the objective into three subsets.

- Objective 1: Loan Performance

In [19]:
loan_performance_columns = [ 'LoanStatus', 'LoanCurrentDaysDelinquent','LoanMonthsSinceOrigination',
    'LoanOriginalAmount', 'BorrowerAPR', 'DebtToIncomeRatio', 'CreditScoreRangeLower','CreditScoreRangeUpper',
    'ProsperScore', 'EmploymentStatus','IsBorrowerHomeowner']

loan_performance_df = df[loan_performance_columns]

- Objective 2: Credit Score and Borrower Analysis

In [20]:
credit_borrower_columns = [ 'CreditScoreRangeLower', 'CreditScoreRangeUpper', 'ProsperRating (numeric)',
    'ProsperRating (Alpha)','ProsperScore','IncomeRange','EmploymentStatus','IsBorrowerHomeowner',
    'BorrowerAPR', 'LoanOriginalAmount']

credit_borrower_df = df[credit_borrower_columns]

- Objective 3: Geographic and Demographic Analysis

In [21]:
geo_demo_columns = [ 'BorrowerState', 'Occupation', 'EmploymentStatus', 'IncomeRange',
    'IsBorrowerHomeowner', 'LoanOriginalAmount', 'ProsperRating (Alpha)', 'CreditScoreRangeLower',
    'CreditScoreRangeUpper']

geo_demo_df = df[geo_demo_columns]

### Features for investigation 

> The following features

## Univariate Exploration

> In this section, investigate distributions of individual variables. If you see unusual points or outliers, take a deeper look to clean things up and prepare yourself to look at relationships between variables.

>**Rubric Tip**: Use the "Question-Visualization-Observations" framework  throughout the exploration. This framework involves **asking a question from the data, creating a visualization to find answers, and then recording observations after each visualisation.** 

> **Rubric Tip**: This part (Univariate Exploration) should include at least one histogram, and either a bar chart of count plot.

>**Rubric Tip**: Visualizations should depict the data appropriately so that the plots are easily interpretable. You should choose an appropriate plot type, data encodings, and formatting as needed. The formatting may include setting/adding the title, labels, legend, and comments. Also, do not overplot or incorrectly plot ordinal data.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your data. Make sure the variables that you cover here have been introduced in some fashion in the previous section (univariate exploration).

> **Rubric Tip**: This part (Bivariate Exploration) should include at least one scatter plot, one box plot, and at least one clustered bar chart or heat map.

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

> **Rubric Tip**: This part (Multivariate Exploration) should include at least one Facet Plot, and one Plot Matrix or Scatterplot with multiple encodings.

>**Rubric Tip**: Think carefully about how you encode variables. Choose appropriate color schemes, markers, or even how Facets are chosen. Also, do not overplot or incorrectly plot ordinal data.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

## Conclusions
>You can write a summary of the main findings and reflect on the steps taken during the data exploration.

> **Rubric Tip**: Create a list of summary findings to make it easy to review.

> Remove all Tips mentioned above, before you convert this notebook to PDF/HTML.


> At the end of your report, make sure that you export the notebook as an html file from the `File > Download as... > HTML or PDF` menu. Make sure you keep track of where the exported file goes, so you can put it in the same folder as this notebook for project submission. Also, make sure you remove all of the quote-formatted guide notes like this one before you finish your report!

